### PR TITLE
generic proxy: new kafka codec support

### PR DESCRIPTION
**Pull request by wbpcode**
Commit Message: generic proxy: new kafka codec support
Additional Description:

Added new kafka codec support to generic proxy. Now, we can use generic proxy to proxy kafka traffic and to modify the kafka traffic with generic filter.

Thanks so much for solid work of @adamkotwasinski in the kafka area. It make the kafka codec for generic proxy so easy!!!

Risk Level: low.
Testing: uinit.
Docs Changes: n/a.
Release Notes: n/a.
Platform Specific Features: n/a.

### CODE REVIEW

File: contrib/generic_proxy/filters/network/source/codecs/kafka/config.cc [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+KafkaServerCodec::KafkaServerCodec() : response_encoder_(response_buffer_) {}
+
+void KafkaServerCodec::setCodecCallbacks(GenericProxy::ServerCodecCallbacks& callbacks) {
+  request_callbacks_ = std::make_shared<KafkaRequestCallbacks>(callbacks);
+  request_decoder_ = std::make_shared<NetworkFilters::Kafka::RequestDecoder>(
+      std::vector<NetworkFilters::Kafka::RequestCallbackSharedPtr>{request_callbacks_});
+}
+
+void KafkaServerCodec::decode(Envoy::Buffer::Instance& buffer, bool) {
+  request_buffer_.move(buffer);
```
Comment (soulxu) [inline contrib/generic_proxy/filters/network/source/codecs/kafka/config.cc:19]: why do we need this buffer? since we just parse the data, then drain it.

Comment (soulxu) [inline contrib/generic_proxy/filters/network/source/codecs/kafka/config.cc:19]: can we use the original buffer?

**Comment by soulxu** [comment]
> Added new kafka codec support to generic proxy. Now, we can use generic proxy to proxy kafka traffic and to modify the kafka traffic with generic filter.
> 

Is there any different goal compare to the kafka filter?

File: contrib/generic_proxy/filters/network/source/codecs/kafka/config.cc (hunk 1 above)
Comment (wbpcode) [reply contrib/generic_proxy/filters/network/source/codecs/kafka/config.cc:19]: Yeah. But when we doing the decoding, the callbacks will be called and I am afraid that the original buffer maybe changed by the callbacks in some ways. (Although it's almost impossible, but moving also very cheap.)

If you think this is a block point, I can change it.

**Comment by wbpcode** [comment]
> Is there any different goal compare to the kafka filter?

Now, we have two kafka filter in the contrib. A kafka broker filter and a kafka mesh filter. This is completely different with kafka mesh filter, and is similar with kafka broker filter.

But there are some difference:
- Access log support from generic proxy.
- L7 filter support from generic proxy.
- Request-response binding automatically.

The initial motivation of creating this codec is that we want to rewrite the kafka message but the kafka broker filter has no a extension point to do that. And we also want logging or even tracing support in the future. So, we create this codec.

**Review by soulxu — APPROVED** [review]
LGTM, thanks

File: contrib/generic_proxy/filters/network/source/codecs/kafka/config.cc (hunk 1 above)
Comment (soulxu) [reply contrib/generic_proxy/filters/network/source/codecs/kafka/config.cc:19]: I see it, I think it's fine.

**Comment by soulxu** [comment]
> > Is there any different goal compare to the kafka filter?
> 
> Now, we have two kafka filter in the contrib. A kafka broker filter and a kafka mesh filter. This is completely different with kafka mesh filter, and is similar with kafka broker filter.
> 
> But there are some difference:
> 
> * Access log support from generic proxy.
> * L7 filter support from generic proxy.
> * Request-response binding automatically.
> 
> The initial motivation of creating this codec is that we want to rewrite the kafka message but the kafka broker filter has no a extension point to do that. And we also want logging or even tracing support in the future. So, we create this codec.

sounds cool for me

**Pull request merged**